commit 9c652fb305a3e3d1b4a752bac9bd8a85d15c7de9
Author: Kefu Chai <kchai@redhat.com>
Date:   Thu Dec 24 15:27:41 2020 +0800

    mgr/ActivePyModules.cc: use wrappers for acquiring/releasing GIL
    
    this change is a follow-up of
    0601b31a53a455f0b67c981460d198cb3a97f3de, for couple reasons
    
    - document the guideline for locking when working with python GIL
    - add primitives to extract the patterns for acquiring/releasing
      GIL. so they can be reused.
    
    Signed-off-by: Kefu Chai <kchai@redhat.com>

diff --git a/src/mgr/ClusterState.h b/src/mgr/ClusterState.h
index e0ac3f31761..eeff1f76bd9 100644
--- a/src/mgr/ClusterState.h
+++ b/src/mgr/ClusterState.h
@@ -74,24 +74,24 @@ public:
   }
 
   template<typename Callback, typename...Args>
-  void with_servicemap(Callback&& cb, Args&&...args) const
+  auto with_servicemap(Callback&& cb, Args&&...args) const
   {
     std::lock_guard l(lock);
-    std::forward<Callback>(cb)(servicemap, std::forward<Args>(args)...);
+    return std::forward<Callback>(cb)(servicemap, std::forward<Args>(args)...);
   }
 
   template<typename Callback, typename...Args>
-  void with_fsmap(Callback&& cb, Args&&...args) const
+  auto with_fsmap(Callback&& cb, Args&&...args) const
   {
     std::lock_guard l(lock);
-    std::forward<Callback>(cb)(fsmap, std::forward<Args>(args)...);
+    return std::forward<Callback>(cb)(fsmap, std::forward<Args>(args)...);
   }
 
   template<typename Callback, typename...Args>
-  void with_mgrmap(Callback&& cb, Args&&...args) const
+  auto with_mgrmap(Callback&& cb, Args&&...args) const
   {
     std::lock_guard l(lock);
-    std::forward<Callback>(cb)(mgr_map, std::forward<Args>(args)...);
+    return std::forward<Callback>(cb)(mgr_map, std::forward<Args>(args)...);
   }
 
   template<typename Callback, typename...Args>
@@ -111,11 +111,11 @@ public:
   }
 
   template<typename... Args>
-  void with_monmap(Args &&... args) const
+  auto with_monmap(Args &&... args) const
   {
     std::lock_guard l(lock);
     ceph_assert(monc != nullptr);
-    monc->with_monmap(std::forward<Args>(args)...);
+    return monc->with_monmap(std::forward<Args>(args)...);
   }
 
   template<typename... Args>
@@ -138,17 +138,17 @@ public:
   }
 
   template<typename Callback, typename...Args>
-  void with_health(Callback&& cb, Args&&...args) const
+  auto with_health(Callback&& cb, Args&&...args) const
   {
     std::lock_guard l(lock);
-    std::forward<Callback>(cb)(health_json, std::forward<Args>(args)...);
+    return std::forward<Callback>(cb)(health_json, std::forward<Args>(args)...);
   }
 
   template<typename Callback, typename...Args>
-  void with_mon_status(Callback&& cb, Args&&...args) const
+  auto with_mon_status(Callback&& cb, Args&&...args) const
   {
     std::lock_guard l(lock);
-    std::forward<Callback>(cb)(mon_status_json, std::forward<Args>(args)...);
+    return std::forward<Callback>(cb)(mon_status_json, std::forward<Args>(args)...);
   }
 
   void final_init();
